bindings/pyroot/ROOT.py

```diff
@@ -238,6 +240,73 @@ def _TTree__iter__( self ):
 
 _root.CreateScopeProxy( "TTree" ).__iter__    = _TTree__iter__
 
+# Array interface
+def _proxy__array_interface__(self):
+    getter_array_interface = "_get__array_interface__"
+    if hasattr(self, getter_array_interface):
+        return self._get__array_interface__()
+    else:
+        raise Exception("Class {} does not have method {}.".format(
+            type(self), getter_array_interface))
+
+for pyclass in [
+        "TVectorT<{DTYPE}>",
+        "TMatrixT<{DTYPE}>",
+        "std::vector<{DTYPE}>",
+        "Experimental::VecOps::TVec<{DTYPE}>"
+        ]:
+    if "TVectorT" in pyclass or "TMatrixT" in pyclass:
+        dtypes = ["float", "double"]
+    else:
+        dtypes = ["float", "double", "int", "unsigned int", "long", "unsigned long"]
+    for dtype in dtypes:
+        class_scope = _root.CreateScopeProxy(pyclass.format(DTYPE=dtype))
+        class_scope._proxy__array_interface__ = _proxy__array_interface__
+        class_scope.__array_interface__ = property(class_scope._proxy__array_interface__)
+
+# TTree.AsMatrix functionality
+def _TTreeAsMatrix(self, columns):
+    try:
+        import numpy as np
+    except:
+        raise ImportError("Failed to import numpy during call of TTree.AsMatrix.")
+
+    # Error-handling
+    if self.GetEntries() == 0:
+        raise Exception("Tree has no entries.")
+
+    # Convert columns interable to std.vector("string")
+    columns_vector = _root.std.vector("string")(len(columns))
+    for i, col in enumerate(columns):
+        columns_vector[i] = col
+
+    # Allocate memory for the read-out
+    # NOTE: This has to be done in PyROOT for a proper handling of the reference counting
+    dtype = "double" # TODO: infere the correct data-type for the output array
```

Comment:
so you mean to return a vector from C++?

Reply:
right, this could be done if the `std.vector` would be piped through PyROOT, which would handle the counting correctly. with the current setup, this is not possible (but i'm not too sure about this).

Reply:
alternatively, we'll need a type promotion mechanism made in python, e.g. ordering types with some type of "rank" and retain the one with the highest rank (e.g. double = 10, float = 9 ....)

Reply:
Done in python. It's a quite heavy implementation...

Reply:
Regarding the concerns of Stefan about reference counting if the vector comes from C++:
```cpp
ROOT.gInterpreter.Declare("std::vector<int> funvec() { return std::vector<int>(); }")
v = ROOT.funvec()
print sys.getrefcount(v) 
```
This actually prints `2`

Reply:
Why does it print `2`? It should be `1`? Are we sure that the garbage collection kicks in if the objects runs out of scope?
However, I've checked that `np.asarray(v)` increases the refcount by 1.

Reply:
To me the example I posted looks like a bug in PyROOT, which does not set properly the reference count of the vector returned by the C++ function before passing it to Python. What you do (using ROOT.std.vector) probably follows a different code path which does set correctly the reference count.